File: tests/prism_test.c
// Test harness for prism geometry module

#include "geometry/prism.h"
#include "test_harness.h"
#include <stdio.h>

TEST_RUNNER_BEGIN();

// =================================================================================================
// Test: Prism Creation
// =================================================================================================

void test_prism_create_basic(void) {
  TEST_BEGIN("prism_create_basic");
  Prism p;
  prism_create(200.0f, 200.0f, 100.0f, 60.0f, &p);

  // Check that apex is above base (y increases downward in screen coords)
  float ax, ay, bx, by;
  prism_get_vertex(&p, 0, &ax, &ay); // apex
  prism_get_vertex(&p, 1, &bx, &by); // bottom-right

  // Apex should have smaller y (higher on screen)
  ASSERT_TRUE(ay < by);

  // Bottom vertices should have same y
  float cx, cy;
  prism_get_vertex(&p, 2, &cx, &cy); // bottom-left
  ASSERT_NEAR(by, cy, 0.001f);

  TEST_END();
}

void test_prism_create_centered(void) {
  TEST_BEGIN("prism_create_centered");
  Prism p;
  prism_create(200.0f, 200.0f, 100.0f, 60.0f, &p);

  // Center of prism should be at (200, 200)
  // Centroid = average of vertices
  float ax, ay, bx, by, cx, cy;
  prism_get_vertex(&p, 0, &ax, &ay);
  prism_get_vertex(&p, 1, &bx, &by);
  prism_get_vertex(&p, 2, &cx, &cy);

  float centroid_x = (ax + bx + cx) / 3.0f;
  float centroid_y = (ay + by + cy) / 3.0f;

  ASSERT_NEAR(centroid_x, 200.0f, 0.1f);
  ASSERT_NEAR(centroid_y, 200.0f, 0.1f);

  TEST_END();
}

void test_prism_create_width(void) {
  TEST_BEGIN("prism_create_width");
  Prism p;
  float size = 100.0f;
  prism_create(200.0f, 200.0f, size, 60.0f, &p);

  // Base width should equal size
  float bx, by, cx, cy;
  prism_get_vertex(&p, 1, &bx, &by);
  prism_get_vertex(&p, 2, &cx, &cy);

  float base_width = bx - cx; // right - left
  ASSERT_NEAR(base_width, size, 0.01f);

  TEST_END();
}

void test_prism_create_clamps_angle(void) {
  TEST_BEGIN("prism_create_clamps_angle");

  // Very small angle (should clamp to 1)
  Prism p1;
  prism_create(200.0f, 200.0f, 100.0f, 0.1f, &p1);
  float scale1 = prism_scale(&p1);
  ASSERT_TRUE(scale1 > 0.0f); // Should produce valid prism

  // Very large angle (should clamp to 179)
  Prism p2;
  prism_create(200.0f, 200.0f, 100.0f, 200.0f, &p2);
  float scale2 = prism_scale(&p2);
  ASSERT_TRUE(scale2 > 0.0f); // Should produce valid prism

  TEST_END();
}

// =================================================================================================
// Test: Prism Scale
// =================================================================================================

void test_prism_scale_positive(void) {
  TEST_BEGIN("prism_scale_positive");
  Prism p;
  prism_create(200.0f, 200.0f, 100.0f, 60.0f, &p);

  float scale = prism_scale(&p);
  ASSERT_TRUE(scale > 0.0f);

  TEST_END();
}

void test_prism_scale_proportional(void) {
  TEST_BEGIN("prism_scale_proportional");

  // Larger prism should have larger scale
  Prism small, large;
  prism_create(200.0f, 200.0f, 50.0f, 60.0f, &small);
  prism_create(200.0f, 200.0f, 200.0f, 60.0f, &large);

  float scale_small = prism_scale(&small);
  float scale_large = prism_scale(&large);

  ASSERT_TRUE(scale_large > scale_small);
  // Should scale roughly linearly with size
  ASSERT_NEAR(scale_large / scale_small, 4.0f, 0.5f);

  TEST_END();
}

// =================================================================================================
// Test: Prism Vertex/Edge Access
// =================================================================================================

void test_prism_get_vertex_valid(void) {
  TEST_BEGIN("prism_get_vertex_valid");
  Prism p;
  prism_create(200.0f, 200.0f, 100.0f, 60.0f, &p);

  float x, y;

  // All vertices should be retrievable
  prism_get_vertex(&p, 0, &x, &y);
  ASSERT_NEAR(x, 200.0f, 0.01f); // apex x centered

  prism_get_vertex(&p, 1, &x, &y);
  ASSERT_TRUE(x > 200.0f); // bottom-right

  prism_get_vertex(&p, 2, &x, &y);
  ASSERT_TRUE(x < 200.0f); // bottom-left

  TEST_END();
}

void test_prism_get_vertex_invalid(void) {
  TEST_BEGIN("prism_get_vertex_invalid");
  Prism p;
  prism_create(200.0f, 200.0f, 100.0f, 60.0f, &p);

  float x, y;

  // Invalid indices should return 0
  prism_get_vertex(&p, -1, &x, &y);
  ASSERT_NEAR(x, 0.0f, 0.001f);
  ASSERT_NEAR(y, 0.0f, 0.001f);

  prism_get_vertex(&p, 3, &x, &y);
  ASSERT_NEAR(x, 0.0f, 0.001f);
  ASSERT_NEAR(y, 0.0f, 0.001f);

  TEST_END();
}

void test_prism_get_edge(void) {
  TEST_BEGIN("prism_get_edge");
  Prism p;
  prism_create(200.0f, 200.0f, 100.0f, 60.0f, &p);

  float ax, ay, bx, by;
  float v0x, v0y, v1x, v1y;

  // Edge 0 should go from v0 to v1
  prism_get_edge(&p, 0, &ax, &ay, &bx, &by);
  prism_get_vertex(&p, 0, &v0x, &v0y);
  prism_get_vertex(&p, 1, &v1x, &v1y);

  ASSERT_NEAR(ax, v0x, 0.001f);
  ASSERT_NEAR(ay, v0y, 0.001f);
  ASSERT_NEAR(bx, v1x, 0.001f);
  ASSERT_NEAR(by, v1y, 0.001f);

  TEST_END();
}

// =================================================================================================
// Test: Point-in-Triangle
// =================================================================================================

void test_point_in_triangle_inside(void) {
  TEST_BEGIN("point_in_triangle_inside");

  // Simple triangle (0,0), (10,0), (5,10)
  int inside = point_in_triangle(5.0f, 5.0f, 0.0f, 0.0f, 10.0f, 0.0f, 5.0f, 10.0f);
  ASSERT_TRUE(inside);

  TEST_END();
}

void test_point_in_triangle_outside(void) {
  TEST_BEGIN("point_in_triangle_outside");

  // Point clearly outside
  int inside = point_in_triangle(20.0f, 20.0f, 0.0f, 0.0f, 10.0f, 0.0f, 5.0f, 10.0f);
  ASSERT_FALSE(inside);

  TEST_END();
}

void test_point_in_triangle_on_edge(void) {
  TEST_BEGIN("point_in_triangle_on_edge");

  // Point on edge (should be considered inside or on boundary)
  // Midpoint of edge (0,0)-(10,0)
  int inside = point_in_triangle(5.0f, 0.0f, 0.0f, 0.0f, 10.0f, 0.0f, 5.0f, 10.0f);
  ASSERT_TRUE(inside);

  TEST_END();
}

void test_point_in_triangle_at_vertex(void) {
  TEST_BEGIN("point_in_triangle_at_vertex");

  // Point at vertex
  int inside = point_in_triangle(0.0f, 0.0f, 0.0f, 0.0f, 10.0f, 0.0f, 5.0f, 10.0f);
  ASSERT_TRUE(inside);

  TEST_END();
}

void test_prism_contains_point_center(void) {
  TEST_BEGIN("prism_contains_point_center");
  Prism p;
  prism_create(200.0f, 200.0f, 100.0f, 60.0f, &p);

  // Center should be inside
  int inside = prism_contains_point(&p, 200.0f, 200.0f);
  ASSERT_TRUE(inside);

  TEST_END();
}

void test_prism_contains_point_outside(void) {
  TEST_BEGIN("prism_contains_point_outside");
  Prism p;
  prism_create(200.0f, 200.0f, 100.0f, 60.0f, &p);

  // Point far outside
  int inside = prism_contains_point(&p, 0.0f, 0.0f);
  ASSERT_FALSE(inside);

  TEST_END();
}

// =================================================================================================
// Main
// =================================================================================================

int main(void) {
  printf("Prism geometry tests\n");
  printf("====================\n\n");

  // Prism creation tests
  test_prism_create_basic();
  test_prism_create_centered();
  test_prism_create_width();
  test_prism_create_clamps_angle();

  // Prism scale tests
  test_prism_scale_positive();
  test_prism_scale_proportional();

  // Vertex/edge access tests
  test_prism_get_vertex_valid();
  test_prism_get_vertex_invalid();
  test_prism_get_edge();

  // Point-in-triangle tests
  test_point_in_triangle_inside();
  test_point_in_triangle_outside();
  test_point_in_triangle_on_edge();
  test_point_in_triangle_at_vertex();
  test_prism_contains_point_center();
  test_prism_contains_point_outside();

  TEST_RUNNER_END();
}

File: tests/prism_test.zig
const std = @import("std");
const testing = std.testing;
const lib = @import("lib");

const prism = lib.prism;
const vec2 = lib.vec2;

test "containsPoint inside triangle" {
    const tri = prism.Prism.init(vec2.xy(100, 100), 60);

    // Centroid should be inside
    const cent = tri.centroid();
    try testing.expect(tri.containsPoint(cent[0], cent[1]));
}

test "containsPoint outside triangle" {
    const tri = prism.Prism.init(vec2.xy(100, 100), 60);

    // Points clearly outside
    try testing.expect(!tri.containsPoint(0, 100)); // far left
    try testing.expect(!tri.containsPoint(200, 100)); // far right
    try testing.expect(!tri.containsPoint(100, 0)); // far above
    try testing.expect(!tri.containsPoint(100, 200)); // far below
}

test "containsPoint on edge" {
    const tri = prism.Prism.init(vec2.xy(100, 100), 60);

    // Point on base edge (bottom, between bottom_right and bottom_left)
    const v1 = tri.vertices.get(.bottom_right);
    const v2 = tri.vertices.get(.bottom_left);
    const mid_base_x = (v1[0] + v2[0]) / 2;
    const mid_base_y = (v1[1] + v2[1]) / 2;
    try testing.expect(tri.containsPoint(mid_base_x, mid_base_y));
}

test "scanlineRange returns correct bounds" {
    // Equilateral: h = base * sqrt(3)/2, apex_offset = base * sqrt(3)/3, base_offset = base * sqrt(3)/6
    // With base=60, center=(100,100): apex at ~(100, 65.4), base at y~117.3
    const tri = prism.Prism.init(vec2.xy(100, 100), 60);

    // Middle scanline at y=100 (center)
    const r = tri.scanlineRange(100);
    try testing.expect(r != null);

    // At center, width should be ~2/3 of base = 40, so x from 80 to 120
    try testing.expectApproxEqAbs(r.?.x_min, 80, 1);
    try testing.expectApproxEqAbs(r.?.x_max, 120, 1);
}

test "scanlineRange outside triangle returns null" {
    const tri = prism.Prism.init(vec2.xy(100, 100), 60);

    // Above triangle (apex is at ~65.4)
    try testing.expect(tri.scanlineRange(60) == null);

    // Below triangle (base is at ~117.3)
    try testing.expect(tri.scanlineRange(125) == null);
}

test "scanlineRange at vertices" {
    const tri = prism.Prism.init(vec2.xy(100, 100), 60);
    const sqrt3 = @sqrt(3.0);
    const apex_y = 100.0 - 60.0 * sqrt3 / 3.0;
    const base_y = 100.0 + 60.0 * sqrt3 / 6.0;

    // At top (apex)
    const range_top = tri.scanlineRange(apex_y);
    try testing.expect(range_top != null);
    try testing.expectApproxEqAbs(range_top.?.x_min, 100, 1);
    try testing.expectApproxEqAbs(range_top.?.x_max, 100, 1);

    // At bottom (base)
    const range_bottom = tri.scanlineRange(base_y);
    try testing.expect(range_bottom != null);
    try testing.expectApproxEqAbs(range_bottom.?.x_min, 70, 1);
    try testing.expectApproxEqAbs(range_bottom.?.x_max, 130, 1);
}

test "equilateral creates symmetric triangle" {
    const center = vec2.xy(100, 100);
    const base = 50.0;
    const tri = prism.Prism.init(center, base);

    const cent = tri.centroid();
    try testing.expectApproxEqAbs(cent[0], 100, 1);
    try testing.expectApproxEqAbs(cent[1], 100, 1);

    // Check symmetry: bottom_right and bottom_left should be equidistant from center
    const v1 = tri.vertices.get(.bottom_right);
    const v2 = tri.vertices.get(.bottom_left);
    const d1 = @sqrt((v1[0] - center[0]) * (v1[0] - center[0]) + (v1[1] - center[1]) * (v1[1] - center[1]));
    const d2 = @sqrt((v2[0] - center[0]) * (v2[0] - center[0]) + (v2[1] - center[1]) * (v2[1] - center[1]));
    try testing.expectApproxEqAbs(d1, d2, 0.1);
}

test "minY and maxY" {
    // Equilateral with base=60, center=(100,100)
    // apex_offset = 60 * sqrt(3)/3 ~ 34.64, base_offset = 60 * sqrt(3)/6 ~ 17.32
    const tri = prism.Prism.init(vec2.xy(100, 100), 60);
    const sqrt3 = @sqrt(3.0);

    try testing.expectApproxEqAbs(tri.minY(), 100.0 - 60.0 * sqrt3 / 3.0, 1);
    try testing.expectApproxEqAbs(tri.maxY(), 100.0 + 60.0 * sqrt3 / 6.0, 1);
}

test "vertices returns correct values" {
    // Equilateral: apex_offset = base * sqrt(3)/3, base_offset = base * sqrt(3)/6
    const tri = prism.Prism.init(vec2.xy(100, 100), 60);
    const sqrt3 = @sqrt(3.0);
    const apex_offset = 60.0 * sqrt3 / 3.0;
    const base_offset = 60.0 * sqrt3 / 6.0;

    const v0 = tri.vertices.get(.apex);
    const v1 = tri.vertices.get(.bottom_right);
    const v2 = tri.vertices.get(.bottom_left);

    // apex
    try testing.expectApproxEqAbs(v0[0], 100, 1);
    try testing.expectApproxEqAbs(v0[1], 100 - apex_offset, 1);

    // v1 = bottom-right
    try testing.expectApproxEqAbs(v1[0], 130, 1);
    try testing.expectApproxEqAbs(v1[1], 100 + base_offset, 1);

    // v2 = bottom-left
    try testing.expectApproxEqAbs(v2[0], 70, 1);
    try testing.expectApproxEqAbs(v2[1], 100 + base_offset, 1);
}
